### DISCUSSION TITLE
reading/writing nested objects

### DISCUSSION THREAD

**Discussion by mohammedalasmar** [nlohmann/json#3434] [Q&A] [answered]
Hi,

I've a nested JSON object that I'd like to be able to retrieve and write.

Example: 
`{
   "type":"config-return",
   "userId":"123456789",
   "master":{
      "configs":{
         "mIkjdfuehnduen":{
            "dataRate":3333,
            "fileSize":1000,
            "fileFormat":"mp4"
         }
      }
   }
}`

I can easily access userId: 
`std::string userId    = rcvdJSON["userId"];`

but how can I access these values dataRate, fileSize and fileFormat? Also, how can I write a JSON object that would look like the example above? 
<img width="329" alt="Screenshot 2022-04-10 at 22 02 32" src="https://user-images.githubusercontent.com/25755632/162639711-e85ea148-890d-4049-b9da-56189359f161.png">

Thanks! 
`

**Comment by falbrechtskirchinger** [comment, marked as answer]
> but how can I access these values dataRate, fileSize and fileFormat?

`json`, the default template instantiation of `basic_json`, uses the STL comtainers `std::map` for objects and `std::vector` for arrays. You can access elements accordingly. For instance:
`int dataRate = rcvdJSON["master"]["configs"]["mIkjdfuehnduen"]["dataRate"];`

I believe the examples section in the README covers everything you're asking:
https://github.com/nlohmann/json#examples

**Reply by mohammedalasmar** [reply]
Thanks for your prompt reply! 
The examples are very useful but I'm still confused about how to retrieve the data from this JSON object. 
In my example above, I need to first read this value:`mIkjdfuehnduen` (it's unknown to me) then I need to use this value to access the other parameters. This is what I'm trying: 

   ` json jObj  = rcvdJSON["master"]["configs"];`

but then how can I find out this value `mIkjdfuehnduen` in `jObj`? also, how can I use `std::vector` or `std::map` here? 

Many thanks!

**Reply by falbrechtskirchinger** [reply]
> Thanks for your prompt reply! The examples are very useful but I'm still confused about how to retrieve the data from this JSON object. In my example above, I need to first read this value:`mIkjdfuehnduen` (it's unknown to me) then I need to use this value to access the other parameters. This is what I'm trying:
> 
> ` json jObj = rcvdJSON["master"]["configs"];`
> 
> but then how can I find out this value `mIkjdfuehnduen` in `jObj`?

Well, if `jObj.is_object()` is `true`, then you can iterate over its children like this:
```cpp
for (auto it = jObj.begin(); it != jObj.end(); ++it) {
    // access object keys using it.key() and values using it.value()
    std::cout << "Key: " << it.key() << "\n";
    std::cout << "Value: " << it.value() << "\n";
}
```

**Reply by mohammedalasmar** [reply]
Perfect! Many thanks!
